lisa/tests/staging/utilclamp.py

```diff
@@ -0,0 +1,311 @@
+                                axis=1
+                            )
+
+        df.set_index('Time', inplace=True)
+
+        return (
+                series_mean(df['capacity']),
+                df['capacity_max'].max(),
+                df.cpu.unique()[-1]
+        )
+
+    @memoized
+    def _get_test_df(self):
+        task = self.rtapp_task_ids_map['task'][0]
+
+        df = self.trace.analysis.rta.df_phases(task)
+        df = df.reset_index()
+        df.rename(columns={'index': 'start'}, inplace=True)
+        df['end'] = df['start'].shift(-1)
+
+        # TODO: To remove once we have named phases.
+        df = df[df.phase > 0]
+
+        df[['capacity', 'capacity_max', 'cpu']] = df.apply(
+                lambda x:
+                self._get_task_activation_mean_capacities(
+                    task, x['start'], x['end']),
+                axis=1, result_type='expand')
+
+        phases = [(0, 0, 0)] + self._get_phases(self.plat_info)
+
+        df[['uclamp_val', 'util', 'band']] = df.apply(
+                lambda x: [phases[int(x.phase)][0],
+                           phases[int(x.phase)][1],
+                           phases[int(x.phase)][2]],
+                axis=1, result_type='expand')
+
+        return df
+
+    @classmethod
+    def _test_passed(cls, df, col, minmax_func):
+        minmax_df = df.apply(minmax_func, axis=1, result_type='expand')
+        passed = df[col].between(minmax_df[0], minmax_df[1])
+
+        return ResultBundle.from_bool(passed.all())
+
+    @FrequencyAnalysis.df_cpus_frequency.used_events
+    @LoadTrackingAnalysis.df_tasks_signal.used_events
+    def test_placement(self) -> ResultBundle:
+        """
+        For each phase, checks if the task placement is compatible with
+        UtilClamp requirements. This is done by comparing the maximum capacity
+        of the CPU on which the task has been placed, with the UtilClamp
+        value.
+        """
+
+        df = self._get_test_df()
+        capacity_margin = self.CAPACITY_MARGIN
+        max_capacities = sorted(set(
+            self.plat_info['cpu-capacities']['rtapp'].values()))
+
+        def expected_minmax(row):
+            uclamp_val = row['uclamp_val']
+            expected = 1024
+
+            for cap in max_capacities:
+                if (cap * capacity_margin) > uclamp_val:
+                    expected = cap
+                    break
+
+            return (expected, expected)
+
+        return self._test_passed(df, 'capacity_max', expected_minmax)
+
+    @FrequencyAnalysis.df_cpus_frequency.used_events
+    @LoadTrackingAnalysis.df_tasks_signal.used_events
+    def test_freq_selection(self) -> ResultBundle:
+        """
+        For each phase, checks if the task placement and frequency selection
+        is compatible with UtilClamp requirements. This is done by comparing
+        the current CPU capacity on which the task has been placed, with the
+        UtilClamp value.
+
+        The expected capacity is :
```

Comment:
Extra space before `:`
Also since there is no blank line after `:`, this becomes a definition list, rather than just a bullet list:
https://docutils.sourceforge.io/docs/user/rst/quickref.html#definition-lists

The expected formatting is:
![image](https://user-images.githubusercontent.com/23336875/101054190-89dd4d80-3580-11eb-908a-234fd50ab2d4.png)
But the one we get from the definition list is:
![image](https://user-images.githubusercontent.com/23336875/101054242-95c90f80-3580-11eb-81e5-9c5634aa01e5.png)


Should be reformatted as:
```
        The expected capacity is:

          - min: The first capacity below the UtilClamp value.
          - max: The estimated selected schedutil capacity.
```

Reply:
done